<center><font size='20' color='blue'>Random Forest</font></center>

In [2]:
import numpy as np
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.metrics import confusion_matrix

In [3]:
df = pd.read_csv(
    '../Data/cancer.csv'
)

In [4]:
df = df.loc[:, ['diagnosis', 'radius_worst', 'texture_worst', 'perimeter_worst', 'area_worst', 'smoothness_worst', 'compactness_worst', 'concavity_worst', 'concave points_worst', 'symmetry_worst', 'fractal_dimension_worst']]

In [7]:
df_m = df[df['diagnosis'] == 'M']
df_b = df[df['diagnosis'] == 'B']

In [8]:
df_b_sample = df_b.sample(212, random_state=32)

In [9]:
df = pd.concat([df_m, df_b_sample], axis=0)

In [12]:
x = df.drop('diagnosis', axis=1)
y = df['diagnosis']

In [13]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=32)

In [16]:
clf = RandomForestClassifier(max_depth=2, random_state=0, criterion='gini').fit(x_train, y_train)

In [17]:
scores = cross_val_score(clf, x_train, y_train, cv=5)

In [34]:
prediction = clf.predict(x_test)

In [35]:
confusion_matrix(y_test, prediction)

array([[45,  4],
       [ 3, 33]])

In [21]:
scores.mean()

0.9469271290605794

In [22]:
scores.std()

0.025596577897466093

In [27]:
scores.var()

0.0006551848000610498

In [19]:
clf_2 = RandomForestClassifier(max_depth=2, random_state=0, criterion='entropy').fit(x_train, y_train)

In [20]:
scores_2 = cross_val_score(clf_2, x_train, y_train, cv=5)

In [31]:
prediction_2 = clf_2.predict(x_test)

In [33]:
confusion_matrix(y_test, prediction_2)

array([[45,  4],
       [ 3, 33]])

In [23]:
scores_2.mean()

0.9439420544337137

In [24]:
scores_2.std()

0.0171835350243644

In [30]:
scores_2.var()

0.000295273875933558

<font size='15' color='yellow'>Analisis del mejor modelo</font>

Gini: precision 0.94 - desviacion 0.025
Entropy: precision 0.94 - desviacion 0.017

Como se puede apreciar el mejor modelo es el de Entropy, con una precision del 94% y una desviacion de 0.017. Ambos modelo estan bastante cercanos entre si y con una precision bastante alta, el factor determinante para elegir el modelo de Entropy es que tiene una desviacion menor que el modelo de Gini. En mi opinión personal, cualquiera de los dos estaría bien para este caso, valores con tan poca diferencia realmente no son deterministicos para hacer una gran diferencia en el modelo, sin mencionar que los valores de la matriz de confusion son iguales.